File: Java/472.concatenated-words.java
/*
 * @lc app=leetcode id=472 lang=java
 *
 * [472] Concatenated Words
 */

// @lc code=start
class Solution {
    /*
    trie + dfs, dfs递归用isword + sub dfs返回boolean
    time: build tree: O(n * len of word), query: O(N * 2^L * L)
    as for query time complexity: O(L) to loop through word, in sub dfs: 
    time = O(L-1) + O(L-2) + ... + O(1) = 2(O(L-2) + O(L-3) + ... + O(1)) = 2(2(O(L-3) + ... + O(1))) = 2^L
    */
    class TrieNode {
        boolean isWord;
        TrieNode[] children;
        public TrieNode() {
            isWord = false;
            children = new TrieNode[26];
        }
    }
    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        List<String> res = new ArrayList<>();
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode cur = root;
            for (char c : word.toCharArray()) {
                if (cur.children[c - 'a'] == null) {
                    cur.children[c - 'a'] = new TrieNode();
                }
                cur = cur.children[c - 'a'];
            }
            cur.isWord = true;
        }
        for (String word : words) {
            if (dfs(word, root, 0, 0)) {
                res.add(word);
            }
        }
        return res;
    }

    private boolean dfs(String word, TrieNode root, int count , int pos) {
        if (pos == word.length()) {
            return count > 1;
        }
        TrieNode cur = root;
        for (int i = pos; i < word.length(); ++i) {
            if (cur.children[word.charAt(i) - 'a'] != null) {
                if (cur.children[word.charAt(i) - 'a'].isWord && dfs(word, root, count + 1, i + 1)) {
                    return true;
                }
                cur = cur.children[word.charAt(i) - 'a'];
            } else return false;
        }
        return false;
    }
}
// @lc code=end

File: Cpp/472.concatenated-words.cpp
/*
 * @lc app=leetcode id=472 lang=cpp
 *
 * [472] Concatenated Words
 */

// @lc code=start
class Solution {
/*
dfs + memo
time without memo: O(N*L^2 * 2^L)
*/

public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        vector<string> res;
        unordered_set<string> dict(words.begin(), words.end());
        for (string word : words) {
            if (dfs(dict, word, 0)) res.push_back(word);
        }
        return res;
    }

private:
    bool dfs(unordered_set<string>& dict, string& word, int count) {
        if (dict.count(word) && count > 1) return true;
        // only break word in the middle
        for (int i = 1; i < word.size(); ++i) {
            string left = word.substr(0, i);
            if (!dict.count(left)) continue;
            string right = word.substr(i);
            if (dict.count(right) || dfs(dict, right, count + 1)) return true;
        }
        return false;
    }
};

/*
optimize: dfs + memo
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        vector<string> res;
        unordered_set<string> dict(words.begin(), words.end());
        unordered_map<string, bool> memo;
        for (string word : words) {
            if (dfs(dict, word, memo)) res.push_back(word);
        }
        return res;
    }

private:
    bool dfs(unordered_set<string>& dict, string& word, unordered_map<string, bool>& memo) {
        if (memo.count(word)) return memo.at(word);
        // only break word in the middle
        for (int i = 1; i < word.size(); ++i) {
            if (dict.count(word.substr(0, i))) {
                string right = word.substr(i);
                if (dict.count(right) || dfs(dict, right, memo)) {
                    memo[word] = true;
                    return true;
                }
            }
        }
        memo[word] = false;
        return false;
    }
};
*/
// @lc code=end

